storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1764 @@ buf_block_t *fsp_get_latched_xdes_page(
+      goto err_exit;
+
+    auto parent_it= m_parent_pages.find(cur_page_no);
+    if (parent_it == m_parent_pages.end())
+    {
+      err= DB_CORRUPTION;
+      goto err_exit;
+    }
+
+    uint32_t parent_page_no= uint32_t(parent_it->second);
+
+    uint32_t parent_offset= uint32_t(parent_it->second >> 32);
+
+    if (parent_offset > srv_page_size - FIL_PAGE_DATA_END)
+    {
+      err= DB_CORRUPTION;
+      goto err_exit;
+    }
+
+    pageAllocator allocator(header_block, iblock, inode, extent_size,
+                            mtr);
+
+    associatedPages related_pages(block, mtr);
+    err= allocator.prepare(new_extent, old_state == XDES_FSEG);
+
+    DBUG_EXECUTE_IF("allocation_prepare_fail", err= DB_CORRUPTION;);
+    if (err) goto err_exit;
+
+    err= related_pages.prepare(parent_page_no);
+    DBUG_EXECUTE_IF("relation_page_prepare_fail", err= DB_CORRUPTION;);
+
+    if (err) goto err_exit;
+
+    allocator.complete();
+
+    /* After allocating the new page, try to prepare the steps
+    of page removal function. Because there is a possiblity that
+    last block in FSEG_NOT_FULL/FSP_FREE_FRAG/FSP_FREE last block
+    could've changed while allocating the new block. */
+    pageRemover remover(header_block, iblock, inode, extent_size,
+                        cur_page_no, mtr);
+    err= remover.prepare();
+ 
+    DBUG_EXECUTE_IF("remover_prepare_fail", err= DB_CORRUPTION;);
+    if (err)
+    {
+      allocator.restore_old_pages();
+      mtr->discard_modifications();
+      goto err_exit;
+    }
+
+    /* Copy the data from old block to new block */
+    buf_block_t *new_block= allocator.get_new_block();
+    uint32_t new_page_no= new_block->page.id().page_no();
+    /* Copy FIL_PAGE_PREV, FIL_PAGE_NEXT */
+    mtr->memcpy(*new_block, new_block->page.frame + FIL_PAGE_PREV,
+                block->page.frame + FIL_PAGE_PREV, 8);
+    mtr->memcpy(*new_block, new_block->page.frame + FIL_PAGE_TYPE,
+                block->page.frame + FIL_PAGE_TYPE,
+		srv_page_size - FIL_PAGE_TYPE - 8);
+
+    /* Assign the new block page number in left, right
+    and parent block */
+    related_pages.complete(new_page_no, parent_offset);
+
+    /* Complete the page free operation */
+    remover.complete(); 
+    /* Add the new page in inode fragment array */
+    allocator.assign_frag_slot();
+
+    if (level)
+    {
+      err= store_child_pages(new_block);
+      if (err) goto err_exit;
+    }
+    goto fetch_next_page;
+  }
+
+  ut_ad(!fsp_tablespace_validate(space));
+  ut_ad(!fseg_validate(space, m_index));
+  root->page.unfix();
+  return DB_SUCCESS;
+}
+
+dberr_t indexDefragmenter::defragment(spaceDefragmenter *space_defrag) noexcept
+{
+  mtr_t mtr;
+  dberr_t err= DB_SUCCESS;
+  m_index->lock.x_lock(SRW_LOCK_CALL);
+  fil_space_t *const space= fil_system.sys_space;
+  mtr.start();
+  mtr.x_lock_space(space);
+  buf_block_t *root= fsp_get_latched_page(page_id_t{0, m_index->page},
+                                          &mtr, &err);
```

Comment:
It does not seem helpful to invoke `fsp_get_latched_page()` instead of `buf_page_get_gen()` at the start of a mini-transaction.

`btr_root_block_get()` would perform more consistency checks. I think that we should invoke that function here. If the root page is valid, we should `fix()` it and `mtr->release_last_page()`, and at the end of this function, invoke `root->page.unfix()`.

There does not seem to be a need to acquire a tablespace latch before acquiring the root page latch. I’d let `defragment_level()` to keep track of that.